### PR TITLE
Add hooks to get firestore document data once

### PR DESCRIPTION
**Pull request by davepwsmith**
Had some spare time and thought I would (try to) contribute instead of just moaning on the issues :-) 

This is part of the way to #137 

Not sure if this is the most efficient way to do this, but it fits in well with the general pattern. 

Presume you will need me to document it (although it will mostly be a copy and paste I should think!!) - I'll do that next week.

### CODE REVIEW

**Comment by jhuleatt** [comment]
@davepwsmith thank you for reviving #137 and for the thorough PR with tests and sample code!

My main question is whether we instead want to handle this in `ReactFireOptions` with a `once` param or similar instead of writing out dedicated `*Once` functions. I'm meeting with @davideast and @jamesdaniels later today, so I'll run this by them too. Any thoughts on the tradeoffs?

**Comment by davepwsmith** [comment]
I don't really see that there is a meaningful difference between those two approaches, from the point of view of a user of the library (of which I am one!) - I would point out that the names of the hooks are getting pretty long! I guess it _could_ be one less thing to import, but I haven't often found that I am using multiple hooks in one component where I can avoid it - not least because #202 precludes passing data between them easily.

I also wanted to take a look at doing the same for collections, but somehow I wasn't sure it seemed as simple. I was using the sample code partly as a way of working out when the observables were returning (the hooks seem to return a value more than once which is interesting... I guess that is the component getting re-rendered, I didn't look into it too hard) and when I tried the same approach with collections on the face of it, it seemed to return the observable every time the collection was modified - still with the first value, but nonetheless, not "Once". I might take another look at that and check that it isn't the sample code itself that is causing that though. 

Let me know any further thoughts.

File: reactfire/firestore/index.tsx [hunk 1]
```diff
@@ -49,6 +50,23 @@ export function useFirestoreDoc<T = unknown>(
...
+ *
+ * @param ref - Reference to the document you want to get
+ * @param options
+ */
+export function useFirestoreDocOnce<T = unknown>(
+  ref: firestore.DocumentReference,
+  options?: ReactFireOptions<T>
+): T extends {} ? T : firestore.DocumentSnapshot {
+  return useObservable(
+    doc(ref).pipe(first()),
```
Comment (jhuleatt) [inline reactfire/firestore/index.tsx]: todo: @davideast mentioned that there may be a problem with `first` never unsubscribing if the document doesn't exist. But not 100% sure

Comment (davepwsmith) [reply reactfire/firestore/index.tsx]: I looked this up and I _think_ that first (as opposed to take(1)) throws an error if the stream completes with no values - https://swalsh.org/blog/rxjs-first-vs-take1 - but I'm a little new to observables, so I could be wrong. Might be something that I could write into a test!

Comment (davepwsmith) [reply reactfire/firestore/index.tsx]: Just added another test (for useFirestoreDocOnce, since rxfire returns empty objects for docData) using a non-existent doc reference... seeeems to work as intended, but I am not sure how I could get inside the hooks and work out if the observable is actually unsubscribed?

See commit below.

Comment (jhuleatt) [reply reactfire/firestore/index.tsx]: Makes sense to me! We were trying to remember whether `take(1)` or `first` was the problematic one. Thanks for following up!

File: reactfire/firestore/index.tsx [hunk 2]
```diff
@@ -49,6 +50,23 @@ export function useFirestoreDoc<T = unknown>(
...
+    'firestore doconce: ' + ref.path,
```
Comment (jhuleatt) [inline reactfire/firestore/index.tsx]: To match the incoming #210, please change this to:

```ts
`useFirestoreDocOnce:${ref.path}:${JSON.stringify(options)}`
```

File: reactfire/firestore/index.tsx [hunk 3]
```diff
@@ -66,6 +84,23 @@ export function useFirestoreDocData<T = unknown>(
...
+ * @param ref - Reference to the document you want to get
+ * @param options
+ */
+export function useFirestoreDocDataOnce<T = unknown>(
+  ref: firestore.DocumentReference,
+  options?: ReactFireOptions<T>
+): T {
+  return useObservable(
+    docData(ref, checkIdField(options)).pipe(first()),
+    'firestore docdataonce: ' + ref.path,
```
Comment (jhuleatt) [inline reactfire/firestore/index.tsx]: To match the incoming #210, please change this to:

```ts
`useFirestoreDocDataOnce:${ref.path}:${JSON.stringify(options)}`
```

File: reactfire/firestore/index.tsx [hunk 4]
```diff
@@ -49,6 +50,23 @@ export function useFirestoreDoc<T = unknown>(
...
+ * @param ref - Reference to the document you want to get
+ * @param options
+ */
+export function useFirestoreDocOnce<T = unknown>(
+  ref: firestore.DocumentReference,
+  options?: ReactFireOptions<T>
+): T extends {} ? T : firestore.DocumentSnapshot {
+  return useObservable(
+    doc(ref).pipe(first()),
+    `useFirestoreDocOnce:${ref.path}:${JSON.stringify(options)}`,
```
Comment (jamesdaniels) [inline reactfire/firestore/index.tsx]: Sorry #210 is a bit of a moving target. Updated the pattern again, let's switch this to `firestore:docOnce:${ref.firestore.app.name}:${ref.path}` and the other one to `firestore:docDataOnce:${ref.firestore.app.name}:${ref.path}:idField=${idField}`. That way we're ensuring we're caching the observable correctly. I'm going to write a util function that does this automatically in the future.

Comment (davepwsmith) [reply reactfire/firestore/index.tsx]: Would you like me to update all the functions as above, or just the ones I have added?

Comment (davepwsmith) [reply reactfire/firestore/index.tsx]: Also, presumably ${idField} may not exist?

Comment (jamesdaniels) [reply reactfire/firestore/index.tsx]: Yeah that won't be a problem, I just want to make sure that if the developer doesn't get unexpected results from it using a prior cached observable when they change `idField`.

**Comment by jamesdaniels** [comment]
Thanks for the contribution @davepwsmith! Look for it in the next release.
